File: ejercicio4/filterChain.rb
class FilterChain
  def initialize(target)
    @filters = []
    @target = target
  end

  def set_target(target)
    @target = target
  end

  def add_filter(filter)
    @filters << filter
  end

  def ejecutar(mensaje)
    cumplimiento = true
    @filters.each do |filter|
      result = filter.ejecutar(mensaje)

      if result == false
        puts "Filtro #{filter.class} ha bloqueado la ejecución. Autenticación Fallida."
        return
      end
    end
    @target.ejecutar(mensaje)
  end
end

File: P2/Ejercicio1/lib/filterChain.dart
import 'struct_credenciales.dart';
import 'filter.dart';
import 'target.dart';
import 'sensorFiltro.dart';

class FilterChain{
  List<Filter> _filters = [];
  Target _target;
  SensorFiltro _sensor;

  FilterChain(this._target, this._sensor);

  void setTarget(Target target){
    _target = target;
  }

  void addFilter(Filter filter){
    _filters.add(filter);
  }
  
  @override
  void ejecutar(Credenciales mensaje){

    for(int i = 0; i < _filters.length; i++){
      bool result = _filters[i].ejecutar(mensaje);

      if(!result){
        _sensor.setError(_filters[i].msgerror());
        return;
      }
    }
    
    _target.ejecutar(mensaje);
  }
}